### DISCUSSION TITLE
Microsoft.UI.Xaml.Input or Windows.UI.Xaml.Input

### DISCUSSION THREAD

**Discussion by HowardPWeiss** [unoplatform/uno#12633] [Q&A]
I am using Uno to create an application to run on both Linux (Visual Studio Code) and Windows (Visual Studio 2022).  The application is written in C#

I am having the following problem:

On Linux, 

"Using Microsoft.UI.Xaml.Input ;" compiles

and

"Using Windows.UI.Xaml.Input;"  will not compile (missing namespace)

On Windows, the opposite is true

How do I configure my solutions so that the same Using statement works compiles on both Linux and Windows

Howard Weiss

**Comment by jeromelaban** [comment]
Thanks for the report! Could you build you project on windows and linux using [build logging](https://platform.uno/docs/articles/uno-builds-troubleshooting.html), so we can troubleshoot this further?

**Comment by HowardPWeiss** [comment]
Jerome -

I am now thinking that this is not a bug

I was building the .Skia.GTK project on Linux and the .Windows project on Windows.  I am using the latest version of VisualStudio2022. The VisualStudio2022 .Windows project specifies

 <OutputType>WinExec</OutputType>
 <TargetFrameworks>net7.0-windows,10.0.19041.0</TArgetFRameworks>

net7.0-windows.10.0.19041.0 includes Windows.UI.Xaml.Input and does not include Microsoft.UI.Xaml.Input.  Hence the build fails

However the .,SKia.GTK project does build and works

Howard

**Reply by jeromelaban** [reply]
That's interesting. This is not supposed to happen when using Uno.WinUI packages. If you're able to provide any binlog of the errors, it will be helpful. Thanks!

**Comment by HowardPWeiss** [comment]
Jerome 

I have my solution running on Linux (Ubuntu 20.04)  using Visual Studio Core. 

Problem 1

---------

I tried copying the solution to Windows 11.  I upgraded Visual Studio 2022 to the latest versions.

I was unable to compile my working .Skia.GDK project.  When I compiled the solution on Windows 11, Visual Studio reported the following error

'PointerPoint' does not contain a definition for 'RawPosition' and no accessible extension method 'RawPosition' accepting a first argument of type 'PointerPoint' could be found (are you missing a using directive or an assembly reference?)

Offending code as follows:


using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Input;
using System;
using System.Diagnostics;

    private void ExplainPointer(string eventName, object sender, PointerRoutedEventArgs e)
    {
         Microsoft.UI.Xaml.Input.Pointer ptr = e.Pointer;

         if (ptr.PointerDeviceType == Microsoft.UI.Input.PointerDeviceType.Mouse)            
         {
              
             Microsoft.UI.Input.PointerPoint ptrPt = e.GetCurrentPoint(sender as Microsoft.UI.Xaml.UIElement);
             Debug.Write(String.Format("[{0}, {1}] raw [{2}, {3}] ", ptrPt.Position.X, ptrPt.Position.Y, ptrPt.RawPosition.X, ptrPt.RawPosition.Y));

This compiled without error under Visual Studio Core.  NB: ptrPt.RawPosition is defined in Windows.UI.Input Namespace. 

FYI: When I GoToDefinition on Linux, Visual Studio Code reports


#region Assembly Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null
// Uno.UI.dll
#endregion

using Windows.Devices.Input;
using Windows.Foundation;

namespace Microsoft.UI.Input
{
    public class PointerPoint
    {
        public PointerPoint(Windows.UI.Input.PointerPoint point);

        public uint FrameId { get; }
        public ulong Timestamp { get; }
        public PointerDevice PointerDevice { get; }
        public PointerDeviceType PointerDeviceType { get; }
        public uint PointerId { get; }
        public Point RawPosition { get; }
        public Point Position { get; }
        public bool IsInContact { get; }
        public PointerPointProperties Properties { get; }

        public override string ToString();

        public static explicit operator Windows.UI.Input.PointerPoint(PointerPoint muxPointerPoint);
    }
}

When I GoToDefinition on Windows 11, Visual Studio 2022 reports

#region Assembly Microsoft.InteractiveExperiences.Projection, Version=10.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
// C:\Users\Owner\.nuget\packages\microsoft.windowsappsdk\1.1.3\lib\net5.0-windows10.0.18362.0\Microsoft.InteractiveExperiences.Projection.dll
// Decompiled with ICSharpCode.Decompiler 7.1.0.6543
#endregion

using System;
using System.Collections.Concurrent;
using System.Runtime.InteropServices;
using System.Threading;
using ABI.Microsoft.UI.Input;
using Microsoft.Foundation;
using Windows.Foundation;
using Windows.Foundation.Metadata;
using WinRT;

namespace Microsoft.UI.Input
{
    [WindowsRuntimeType("Microsoft.UI")]
    [ProjectedRuntimeClass("_default")]
    [ObjectReferenceWrapper("_inner")]
    [ContractVersion(typeof(WindowsAppSDKContract), 65536u)]
    public sealed class PointerPoint : ICustomQueryInterface, IWinRTObject, IDynamicInterfaceCastable, IEquatable<PointerPoint>
    {
        [StructLayout(LayoutKind.Sequential, Size = 1)]
        private struct InterfaceTag<I>
        {
        }

        private IObjectReference _inner;

        private volatile ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> _queryInterfaceCache;

        private volatile ConcurrentDictionary<RuntimeTypeHandle, object> _additionalTypeData;

        private IntPtr ThisPtr
        {
            get
            {
                if (_inner != null)
                {
                    return _inner.ThisPtr;
                }

                return ((IWinRTObject)this).NativeObject.ThisPtr;
            }
        }

        private IObjectReference _objRef_global__Microsoft_UI_Input_IPointerPoint => _inner;

        private IPointerPoint _default => null;

        bool IWinRTObject.HasUnwrappableNativeObject => true;

        IObjectReference IWinRTObject.NativeObject => _inner;

        ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> IWinRTObject.QueryInterfaceCache => _queryInterfaceCache ?? MakeQueryInterfaceCache();

        ConcurrentDictionary<RuntimeTypeHandle, object> IWinRTObject.AdditionalTypeData => _additionalTypeData ?? MakeAdditionalTypeData();

        public uint FrameId => IPointerPointMethods.get_FrameId(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public bool IsInContact => IPointerPointMethods.get_IsInContact(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public PointerDeviceType PointerDeviceType => IPointerPointMethods.get_PointerDeviceType(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public uint PointerId => IPointerPointMethods.get_PointerId(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public Point Position => IPointerPointMethods.get_Position(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public PointerPointProperties Properties => IPointerPointMethods.get_Properties(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public ulong Timestamp => IPointerPointMethods.get_Timestamp(_objRef_global__Microsoft_UI_Input_IPointerPoint);

        public static PointerPoint FromAbi(IntPtr thisPtr)
        {
            if (thisPtr == IntPtr.Zero)
            {
                return null;
            }

            return MarshalInspectable<PointerPoint>.FromAbi(thisPtr);
        }

        internal PointerPoint(IObjectReference objRef)
        {
            _inner = objRef.As(GuidGenerator.GetIID(typeof(IPointerPoint).GetHelperType()));
        }

        public static bool operator ==(PointerPoint x, PointerPoint y)
        {
            return (x?.ThisPtr ?? IntPtr.Zero) == (y?.ThisPtr ?? IntPtr.Zero);
        }

        public static bool operator !=(PointerPoint x, PointerPoint y)
        {
            return !(x == y);
        }

        public bool Equals(PointerPoint other)
        {
            return this == other;
        }

        public override bool Equals(object obj)
        {
            PointerPoint pointerPoint = obj as PointerPoint;
            if ((object)pointerPoint != null)
            {
                return this == pointerPoint;
            }

            return false;
        }

        public override int GetHashCode()
        {
            return ThisPtr.GetHashCode();
        }

        private ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> MakeQueryInterfaceCache()
        {
            Interlocked.CompareExchange(ref _queryInterfaceCache, new ConcurrentDictionary<RuntimeTypeHandle, IObjectReference>(), null);
            return _queryInterfaceCache;
        }

        private ConcurrentDictionary<RuntimeTypeHandle, object> MakeAdditionalTypeData()
        {
            Interlocked.CompareExchange(ref _additionalTypeData, new ConcurrentDictionary<RuntimeTypeHandle, object>(), null);
            return _additionalTypeData;
        }

        public PointerPoint GetTransformedPoint(IPointerPointTransform transform)
        {
            return IPointerPointMethods.GetTransformedPoint(_objRef_global__Microsoft_UI_Input_IPointerPoint, transform);
        }

        private bool IsOverridableInterface(Guid iid)
        {
            return false;
        }

        CustomQueryInterfaceResult ICustomQueryInterface.GetInterface(ref Guid iid, out IntPtr ppv)
        {
            ppv = IntPtr.Zero;
            if (IsOverridableInterface(iid) || WinRT.InterfaceIIDs.IInspectable_IID == iid)
            {
                return CustomQueryInterfaceResult.NotHandled;
            }

            if (((IWinRTObject)this).NativeObject.TryAs(iid, out ppv) >= 0)
            {
                return CustomQueryInterfaceResult.Handled;
            }

            return CustomQueryInterfaceResult.NotHandled;
        }
    }
}
#if false // Decompilation log


RawPosition is defined in Linux (Assembly Uno.UI) but not in Windows (Microsoftr.InteractiveExperiences.Projection). 

Problem 2
---------

I created a new solution under VisualStudio 2022 and copied the necessary files from the Linux based solution to the new Windows based solution

One of the files I transferred was BoolToVisibilityConverter.cs (a legacy from a previous version) 

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Data;
using Microsoft.UI.Xaml.Input;


namespace SudokuUnoTest
{
    class BoolToVisibilityConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, string language)
        {
            if (!(value is bool))
            {
                return Visibility.Collapsed;
            }
            bool objValue = (bool)value;
            if (objValue)
            {
                return Visibility.Visible;
            }
            return Visibility.Collapsed;
        }

        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            try
            {
                if ((bool)value)
                {
                    return Visibility.Visible;
                }
            }
            catch { }
            return Visibility.Collapsed;
        }
    }
}
using System;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Data;
using Microsoft.UI.Xaml.Input;


namespace SudokuUnoTest
{
    class BoolToVisibilityConverter : IValueConverter
    {
        public object Convert(object value, Type targetType, object parameter, string language)
        {
            if (!(value is bool))
            {
                return Visibility.Collapsed;
            }
            bool objValue = (bool)value;
            if (objValue)
            {
                return Visibility.Visible;
            }
            return Visibility.Collapsed;
        }

        public object ConvertBack(object value, Type targetType, object parameter, string language)
        {
            try
            {
                if ((bool)value)
                {
                    return Visibility.Visible;
                }
            }
            catch { }
            return Visibility.Collapsed;
        }
    }
}


Compiling this file resulted in the following errors

The type or namespace name 'Data' does not exist in the namespace 'Microsoft.UI.Xaml' (are you missing an assembly reference?)
The type or namespace name 'Input' does not exist in the namespace 'Microsoft.UI.Xaml' (are you missing an assembly reference?)
The type or namespace name 'IValueConverter' could not be found (are you missing a using directive or an assembly reference?)

In VisualStudio2022, IValueConverter is defined in System.Windows.Data. To compile this file, I had to modify the source by removing the using Microsoft.UI.Xaml statements and inserting using Systems.Windows and using System.Windows.Data

A similar problem existed for other files which referenced Microsoft.UI.Xaml namepsaces.  Once I substituted the the appropriate Windows namespace, the files compiled

However, there remains one additional block of code which I do not know how to fix.  This references the WinRT methods to instantiae a FilePickerWindow and choose a file

        public async void ReadClick()
        {
            Debug.WriteLine(String.Format("ReadClick"));

            var fileOpenPicker = new FileOpenPicker();
            fileOpenPicker.SuggestedStartLocation = PickerLocationId.ComputerFolder;
            fileOpenPicker.FileTypeFilter.Add(".txt");
            fileOpenPicker.FileTypeFilter.Add(".csv");

            // For Uno.WinUI-based apps
            var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(App.MainWindow);
            WinRT.Interop.InitializeWithWindow.Initialize(fileOpenPicker, hwnd);

            StorageFile pickedFile = await fileOpenPicker.PickSingleFileAsync();

            if (pickedFile != null)
            {
                _worker.btnReset_Click(null, null);
                Debug.WriteLine(String.Format("PIcked name {0} path {1}", pickedFile.Name, pickedFile.Path));
                string[] gameboard = File.ReadAllLines(pickedFile.Path);              
                for (int ii = 0; ii  < gameboard.Length; ii++)
                {
                    Debug.WriteLine(gameboard[ii]);
                }
                ProcessGameboard(gameboard);
            }
            else
            {
                Debug.WriteLine("picked file is null");

            }
        }

The namespace WinRt does not exist.  I am unsure how to instantiate a FilePicker (OpenFileDialog) in a way which works on both Windows and Linux.  I more or less expected that the Uno Platform would provide such a method

Howard Weiss

**Comment by HowardPWeiss** [comment]
Jerome

Please forget the problem with WinRT.  The two statements containing WinRT are unnecessary

Howard
